### DISCUSSION TITLE
different values between forward_train and forward_inferece for contextblockconformer in eval mode

### DISCUSSION THREAD

**Discussion by jeon30c** [espnet/espnet#3971] [Q&A]
```
config_file = 'exp/asr_train_asr_streaming_conformer_transducer_raw_en_bpe500_sp/config.yaml'
model_file = ‘exp/asr_train_asr_streaming_conformer_transducer_raw_en_bpe500_sp/100epoch.pth’
from espnet2.tasks.asr import ASRTask
asr_model, asr_train_args = ASRTask.build_model_from_file(config_file, model_file)
asr_model.eval()
import torch
data_in = torch.randn(1, 3000, 80)
data_len = torch.tensor([3000]).long()
ys_pad_train, olen, prev_states = asr_model.encoder.forward_train(data_in, data_len)
ys_pad_infer, olen, prev_states = asr_model.encoder.forward_infer(data_in, data_len)
```

After loading the streaming model, I compared the outputs from forward_train and forward_infer methods. The encoder is based on the contextual block conformer. 

I understand that the forward_train is implemented for training efficiency and forward_infer is for chunk-wise processing. Even though the encoder has two forward implementations, I guess that the results from them need to be the same.

According to the outputs from the above codes, I got different results from the methods. Can you explain the value difference of the methods?

**Comment by sw005320** [comment]
@eml914 and @D-Keqi, do you have some answers?

**Reply by D-Keqi** [reply]
@jeon30c Hi! I think this may be because mask used for block processing is employed for "forward_train" but not for "forward_infer". The mask is used for achiving high training efficiency. And more details about this mask can be seen in the paper https://arxiv.org/pdf/1910.07204.pdf.

**Reply by eml914** [reply]
Hmm, I think these values should be the same whether the mask is used or not.  I will look into this at my end and try to reproduce the situation as well as look through the codes again.  Thank you for pointing out.

**Reply by jeon30c** [reply]
Hi, @eml914  As I wrote above, I want to compare the results with respect to forward_train and forward_infer.
Now I find that the inference results are all from forward_train rather than forward_infer because abstract asr class does not utilize infer_mode flag (default false) and contextblockencoder's forward method always executes the below line,

https://github.com/espnet/espnet/blob/53e7c1ffa35a149ae0e6e21a7bd8d2c2de77eff3/espnet2/asr/encoder/contextual_block_conformer_encoder.py#L225

I think that using infer_mode flag is not general considering the other encoder structures. So on your way of checking into the codes, please consider this point.
